library/include/ck/library/reference_tensor_operation/cpu/reference_contraction.hpp

```diff
@@ -0,0 +1,191 @@
+
+#pragma once
+
+#include <iostream>
+#include <sstream>
+
+#include "ck/tensor_operation/gpu/device/device_base.hpp"
+#include "ck/library/utility/host_tensor.hpp"
+
+#include "ck/tensor_operation/gpu/element/element_wise_operation.hpp"
+
+using Bilinear    = ck::tensor_operation::element_wise::Bilinear;
+using Scale       = ck::tensor_operation::element_wise::Scale;
+using PassThrough = ck::tensor_operation::element_wise::PassThrough;
+
+namespace ck {
+namespace tensor_operation {
+namespace host {
+
+// hardcoded for NumDimM == NumDimN == NumDimK == 2
+template <ck::index_t NumDimM,
+          ck::index_t NumDimN,
+          ck::index_t NumDimK,
+          typename ADataType,
+          typename BDataType,
+          typename CDataType,
+          typename AccDataType,
+          typename AElementwiseOperation,
+          typename BElementwiseOperation,
+          typename CDEElementwiseOperation,
+          typename DDataType                                                  = float,
+          ck::enable_if_t<NumDimM == 2 && NumDimN == 2 && NumDimK == 2, bool> = false>
+struct ReferenceContraction_M2_N2_K2 : public ck::tensor_operation::device::BaseOperator
+{
+    // Argument
+    struct Argument : public ck::tensor_operation::device::BaseArgument
+    {
+        Argument(const Tensor<ADataType>& a_ms_ks,
+                 const Tensor<BDataType>& b_ns_ks,
+                 const Tensor<DDataType>& d_ms_ns,
+                 Tensor<CDataType>& c_ms_ns,
+                 AElementwiseOperation a_element_op,
+                 BElementwiseOperation b_element_op,
+                 CDEElementwiseOperation cde_element_op)
+            : a_ms_ks_{a_ms_ks},
+              b_ns_ks_{b_ns_ks},
+              d_ms_ns_{d_ms_ns},
+              c_ms_ns_{c_ms_ns},
+              a_element_op_{a_element_op},
+              b_element_op_{b_element_op},
+              cde_element_op_{cde_element_op}
+        {
+        }
+
+        const Tensor<ADataType>& a_ms_ks_;
+        const Tensor<BDataType>& b_ns_ks_;
+        const Tensor<DDataType>& d_ms_ns_;
+        Tensor<CDataType>& c_ms_ns_;
+
+        AElementwiseOperation a_element_op_;
+        BElementwiseOperation b_element_op_;
+        CDEElementwiseOperation cde_element_op_;
+    };
+
+    // Invoker
+    struct Invoker : public ck::tensor_operation::device::BaseInvoker
+    {
+        using Argument = ReferenceContraction_M2_N2_K2::Argument;
+
+        template <typename Op>
+        void apply_op(Op& op, DDataType& v_d, CDataType& v_c, AccDataType& v_acc)
+        {
+            op(v_c, static_cast<AccDataType>(v_d + v_acc));
+        }
+
+        template <>
+        void apply_op<const Bilinear>(const Bilinear& bilinear,
+                                      DDataType& v_d,
+                                      CDataType& v_c,
+                                      AccDataType& v_acc)
+        {
+            bilinear(v_c, v_d, v_acc);
+        }
```

Comment:
This solution is OK, however it's not scalable, generic and hard to maintain. What if you would like to run different binary elementwise operation? I think we should have here a mechanism similar to this one: https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/develop/include/ck/tensor_operation/gpu/thread/threadwise_tensor_slice_transfer_v7.hpp#L150-L180

Reply:
I made ref_contraction generic (without cde_element_op)